File: solutions/2026-08-30/python/2026-08-30-removing-minimum-and-maximum-from-array.py
```python
# This problem asks us to find the minimum number of deletions to remove both the minimum
# and maximum elements from a given array. Deletions can only occur from the front or back
# of the array.
#
# Link: https://leetcode.com/problems/removing-minimum-and-maximum-from-array/
#
# Approach:
# The core idea is to find the indices of the minimum and maximum elements.
# Once we have these indices, we can consider three scenarios to remove both elements:
# 1. Remove both from the left: This involves deleting all elements up to and including
#    the maximum of the two indices. The number of deletions is `max(min_idx, max_idx) + 1`.
# 2. Remove both from the right: This involves deleting all elements from the end of the
#    array up to and including the minimum of the two indices. The number of deletions
#    is `n - min(min_idx, max_idx)`, where `n` is the length of the array.
# 3. Remove one from the left and one from the right: This means we remove elements from
#    the left up to one of the elements, and then from the right up to the other.
#    There are two sub-cases here:
#    a. Remove min from left, max from right: `min_idx + 1` (from left) + `n - max_idx` (from right).
#    b. Remove max from left, min from right: `max_idx + 1` (from left) + `n - min_idx` (from right).
#
#    Alternatively, and more concisely for case 3, we can think of it as removing elements
#    from the left up to the element that is *further* from the left end, and then removing
#    elements from the right up to the element that is *further* from the right end.
#    The total deletions in this scenario would be `max(min_idx, max_idx) + 1` (from left) +
#    `n - min(min_idx, max_idx)` (from right), but this overcounts. A simpler way to express
#    the third strategy is to remove the element further from the front from the front,
#    and the element further from the back from the back. The total deletions would be
#    the index of the element further from the front + 1 (to include that element)
#    plus the number of elements from the back to the element further from the back.
#    Let's say min_idx is at `i` and max_idx is at `j`.
#    - Remove from left only: `max(i, j) + 1`
#    - Remove from right only: `n - min(i, j)`
#    - Remove one from left, one from right:
#      - Min from left, Max from right: `i + 1 + n - j`
#      - Max from left, Min from right: `j + 1 + n - i`
#    The minimum of these three strategies will be our answer.
#
# Time Complexity:
# O(N) to find the minimum and maximum elements and their indices.
# O(1) for calculating the three possible deletion counts and finding the minimum.
# Therefore, the overall time complexity is O(N), where N is the number of elements in the array.
#
# Space Complexity:
# O(1) as we only use a few variables to store indices and the minimum result.

class Solution:
    def removeMinMax(self, nums: list[int]) -> int:
        n = len(nums)

        # Handle the edge case where the array has only one element.
        # This element is both the minimum and maximum, so 1 deletion is needed.
        if n == 1:
            return 1

        # Find the minimum and maximum elements and their indices.
        min_val = float('inf')
        max_val = float('-inf')
        min_idx = -1
        max_idx = -1

        for i in range(n):
            if nums[i] < min_val:
                min_val = nums[i]
                min_idx = i
            if nums[i] > max_val:
                max_val = nums[i]
                max_idx = i

        # Ensure min_idx is always less than or equal to max_idx for simpler logic.
        # This is not strictly necessary but can make comparisons slightly cleaner.
        # However, the logic below correctly handles any order of min_idx and max_idx.

        # Calculate the number of deletions for three main strategies:

        # Strategy 1: Remove both from the left.
        # We need to delete all elements from the start up to the element that is
        # further from the left end. The index of this element is max(min_idx, max_idx).
        # The number of deletions is this index + 1 (to include the element itself).
        deletions_from_left = max(min_idx, max_idx) + 1

        # Strategy 2: Remove both from the right.
        # We need to delete all elements from the end up to the element that is
        # further from the right end. The index of the element further from the
        # right end is min(min_idx, max_idx).
        # The number of elements from the end to this index is n - index.
        # So, deletions = n - min(min_idx, max_idx).
        deletions_from_right = n - min(min_idx, max_idx)

        # Strategy 3: Remove one from the left and one from the right.
        # There are two ways this can happen:
        #   a) Remove min_idx from the left, max_idx from the right.
        #      Deletions = (min_idx + 1) + (n - max_idx)
        #   b) Remove max_idx from the left, min_idx from the right.
        #      Deletions = (max_idx + 1) + (n - min_idx)
        # We want the minimum of these two possibilities for this strategy.
        # A simpler way to think about this combined strategy is to remove the element
        # that is further from the left (say at index `farthest_idx_from_left`) from the left,
        # and the element that is further from the right (say at index `farthest_idx_from_right`)
        # from the right.
        # The total number of deletions is (farthest_idx_from_left + 1) + (n - farthest_idx_from_right).
        # This can be calculated as:
        # (index of the element that is further from the left) + 1
        # + (number of elements from the end up to the element that is further from the right)
        #
        # Example: If min_idx=2, max_idx=7, n=10
        # Case a: min from left, max from right => (2+1) + (10-7) = 3 + 3 = 6
        # Case b: max from left, min from right => (7+1) + (10-2) = 8 + 8 = 16
        # Minimum for this strategy is 6.
        #
        # The expression `(min_idx + 1) + (n - max_idx)` covers one scenario,
        # and `(max_idx + 1) + (n - min_idx)` covers the other.
        # We take the minimum of these two.
        deletions_one_left_one_right = min(
            (min_idx + 1) + (n - max_idx),  # Remove min from left, max from right
            (max_idx + 1) + (n - min_idx)   # Remove max from left, min from right
        )

        # The final answer is the minimum of the three strategies.
        return min(deletions_from_left, deletions_from_right, deletions_one_left_one_right)

```

File: solutions/2026-08-30/javascript/2026-08-30-removing-minimum-and-maximum-from-array.js
// Problem: Removing Minimum and Maximum From Array
// Link: https://leetcode.com/problems/removing-minimum-and-maximum-from-array/
//
// Approach:
// The problem asks for the minimum number of deletions to remove both the minimum and maximum elements from an array.
// Deletions can only be performed from the front or the back of the array.
//
// The core idea is to find the indices of the minimum and maximum elements. Let these indices be `minIndex` and `maxIndex`.
// We need to consider three scenarios for removing both elements:
// 1. Remove both from the left: This involves deleting elements from the beginning up to and including the element furthest from the start (i.e., the one with the larger index). The number of deletions would be `max(minIndex, maxIndex) + 1`.
// 2. Remove both from the right: This involves deleting elements from the end up to and including the element furthest from the end (i.e., the one with the smaller index). The number of deletions would be `n - min(minIndex, maxIndex)`, where `n` is the length of the array.
// 3. Remove one from the left and one from the right: This means deleting elements from the left up to one of the elements, and from the right up to the other element.
//    - If `minIndex` is to the left of `maxIndex`:
//      - Remove `minIndex` from the left (cost: `minIndex + 1`) and `maxIndex` from the right (cost: `n - maxIndex`). Total cost: `(minIndex + 1) + (n - maxIndex)`.
//    - If `maxIndex` is to the left of `minIndex`:
//      - Remove `maxIndex` from the left (cost: `maxIndex + 1`) and `minIndex` from the right (cost: `n - minIndex`). Total cost: `(maxIndex + 1) + (n - minIndex)`.
//    The total cost for this scenario is `(min(minIndex, maxIndex) + 1) + (n - max(minIndex, maxIndex))`.
//
// We can simplify this by realizing that after finding `minIndex` and `maxIndex`, we can always sort them such that `minIndex <= maxIndex`.
// Let `leftmostIndex = min(minIndex, maxIndex)` and `rightmostIndex = max(minIndex, maxIndex)`.
// The three scenarios then become:
// 1. Remove both from the left: `rightmostIndex + 1` deletions.
// 2. Remove both from the right: `n - leftmostIndex` deletions.
// 3. Remove `leftmostIndex` from the left and `rightmostIndex` from the right: `(leftmostIndex + 1) + (n - rightmostIndex)` deletions.
//
// The minimum of these three values will be our answer.
//
// Edge case: If `nums.length` is 1, that single element is both min and max, so 1 deletion is needed.
//
// Time Complexity:
// - Finding the minimum and maximum values and their indices: O(n), where n is the length of the array.
// - Calculating the three possible deletion counts: O(1).
// - Finding the minimum of the three counts: O(1).
// Therefore, the overall time complexity is O(n).
//
// Space Complexity:
// - We only use a few variables to store indices and counts.
// Therefore, the space complexity is O(1).
//

/**
 * @param {number[]} nums
 * @return {number}
 */
var removeMinMax = function(nums) {
    // Handle the edge case where the array has only one element.
    // In this case, that element is both the minimum and maximum, requiring 1 deletion.
    if (nums.length === 1) {
        return 1;
    }

    // Find the minimum and maximum values in the array and their indices.
    let minVal = nums[0];
    let maxVal = nums[0];
    let minIndex = 0;
    let maxIndex = 0;

    // Iterate through the array starting from the second element.
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] < minVal) {
            minVal = nums[i];
            minIndex = i;
        }
        if (nums[i] > maxVal) {
            maxVal = nums[i];
            maxIndex = i;
        }
    }

    // Get the total number of elements in the array.
    const n = nums.length;

    // Ensure minIndex is the smaller index and maxIndex is the larger index for simpler calculation.
    // This is useful for calculating deletions from left and right.
    const leftmostIndex = Math.min(minIndex, maxIndex);
    const rightmostIndex = Math.max(minIndex, maxIndex);

    // Calculate the three possible scenarios for minimum deletions:

    // Scenario 1: Remove both from the left side.
    // This means deleting all elements from the start up to and including the element
    // that is furthest from the left end (which is the one at `rightmostIndex`).
    // The number of deletions is `rightmostIndex + 1` (since indices are 0-based).
    const deleteFromLeftOnly = rightmostIndex + 1;

    // Scenario 2: Remove both from the right side.
    // This means deleting all elements from the end up to and including the element
    // that is furthest from the right end (which is the one at `leftmostIndex`).
    // The number of deletions is `n - leftmostIndex`.
    const deleteFromRightOnly = n - leftmostIndex;

    // Scenario 3: Remove the leftmost element from the left and the rightmost element from the right.
    // - Deleting the element at `leftmostIndex` from the left requires `leftmostIndex + 1` deletions.
    // - Deleting the element at `rightmostIndex` from the right requires `n - rightmostIndex` deletions.
    // The total deletions for this scenario is the sum of these two.
    const deleteFromBothSides = (leftmostIndex + 1) + (n - rightmostIndex);

    // The minimum number of deletions is the minimum of these three scenarios.
    return Math.min(deleteFromLeftOnly, deleteFromRightOnly, deleteFromBothSides);
};
```